File: src/main/java/io/github/intellij/dlanguage/stubs/types/AutoAssignmentStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import io.github.intellij.dlanguage.psi.impl.named.DLanguageAutoAssignmentImpl;
import io.github.intellij.dlanguage.psi.named.DLanguageAutoAssignment;
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes;
import io.github.intellij.dlanguage.stubs.DLanguageAutoAssignmentStub;
import java.io.IOException;
import org.jetbrains.annotations.NotNull;

public class AutoAssignmentStubElementType extends
    DNamedStubElementType<DLanguageAutoAssignmentStub, DLanguageAutoAssignment> {
    public AutoAssignmentStubElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DLanguageAutoAssignment createPsi(@NotNull final DLanguageAutoAssignmentStub stub) {
        return new DLanguageAutoAssignmentImpl(stub, this);
    }

    @NotNull
    @Override
    public DLanguageAutoAssignmentStub createStub(@NotNull final DLanguageAutoAssignment psi,
                                                  final StubElement parentStub) {
        return new DLanguageAutoAssignmentStub(parentStub, this, psi.getName(),
            psi.getAttributes());
    }

    @Override
    public void serialize(@NotNull final DLanguageAutoAssignmentStub stub, @NotNull final StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
        stub.getAttributes().write(dataStream);
    }

    @NotNull
    @Override
    public DLanguageAutoAssignmentStub deserialize(@NotNull final StubInputStream dataStream, final StubElement parentStub) throws IOException {
        return new DLanguageAutoAssignmentStub(parentStub, this, dataStream.readName(),
            DAttributes.Companion.read(dataStream));
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/AutoAssignmentStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.impl.named.DLanguageAutoAssignmentImpl
import io.github.intellij.dlanguage.psi.named.DLanguageAutoAssignment
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes.Companion.read
import io.github.intellij.dlanguage.stubs.DLanguageAutoAssignmentStub
import java.io.IOException

class AutoAssignmentStubElementType(debugName: String) :
    DNamedStubElementType<DLanguageAutoAssignmentStub, DLanguageAutoAssignment>(debugName) {
    override fun createPsi(stub: DLanguageAutoAssignmentStub): DLanguageAutoAssignment {
        return DLanguageAutoAssignmentImpl(stub, this)
    }

    override fun createStub(
        psi: DLanguageAutoAssignment,
        parentStub: StubElement<*>?
    ): DLanguageAutoAssignmentStub {
        return DLanguageAutoAssignmentStub(
            parentStub, this, psi.name,
            psi.attributes
        )
    }

    @Throws(IOException::class)
    override fun serialize(stub: DLanguageAutoAssignmentStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)
        stub.attributes.write(dataStream)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>?): DLanguageAutoAssignmentStub {
        return DLanguageAutoAssignmentStub(
            parentStub, this, dataStream.readName(),
            read(dataStream)
        )
    }
}
